### DISCUSSION TITLE
[@types/lodash] groupBy allows grouping by non-existent fields

### DISCUSSION THREAD

**Discussion by erezarnon** [DefinitelyTyped/DefinitelyTyped#67502] [Issues with a @types package]
The below should give an error on `r` not existing on our type. But it compiles fine.

```
_.groupBy([{a: 2}], 'r')
```

[a comment by an automated account is not shown]

**Comment by jgerstle** [comment]
This is not just groupBy it's a whole bunch of other functions as well (ie keyBy, some)

**Comment by tarczynskitomek** [comment]
We had a bug recently related to the behaviour.

# How it is?
Lodash offers a convenient method on chains to group values by given property. Let's use an example:

```typescript
const users: User[] = [
  { id: 1, type: 'premium' },
  { id: 2, type: 'regular' },
];

const userIdsByType: { [p: string] : string[] } = _.chain(users)
  .groupBy((user: User) => user.type)
  .mapValues((users: User[]) => users.map((u) => u.id))
  .values();
```

Notice that the dictionary type returned by lowdash: `{ [p: string]: string[] }` declares that for every key `p` of type `string` there will be a corresponding value of type `string[]`. The above though, is not true in runtime. Notice that the `users` array may be:
1. empty - then accessing `userIdsByType.premium` and `userIdsByType.regular` will return `undefined`, and not a string[];
2. missing either premium or regular users, then the corresponding property will return `undefined` and not a string[]

## How it should be?
The type returned from `groupBy().values()` should represent the fact, that the dictionary may not contain values for some keys:

```typescript
const userIdsByType: { [p: string] : string[] | undefined } = _.chain(users) // notice the value type is now `string[] | undefined`
  .groupBy((user: User) => user.type)
  .mapValues((users: User[]) => users.map((user: User) => user.id))
  .values();
```

## Example of buggy behaviour
```typescript
const users: User[] = [
  { id: 1, type: 'premium' },
  { id: 2, type: 'regular' },
];

const userIdsByType: { [p: string] : string[] } = _.chain(users) // the type declares that for each key, there will be a value of type string[]
  .groupBy((user: User) => user.type)
  .mapValues((users: User[]) => users.map((u) => u.id))
  .values();

// the function expects two string arrays, and will throw an error given an `undefined` value
const handleUsers = (premiumIds: string[], regularIds: []) => { 
  // do something with ids
}
```

Since the type returned from `groupBy().values()` is `{ [p: string]: string[] }` calling `handleUsers(userIdsByType.premium, userIdsByType.regular)` won't raise a Typscript compile error. In runtime, though, either or both of the properties may be undefined leading to an Error and buggy behaviour.

Using the proposed type: `{ [p: string]: string[] | undefined }` describes the runtime more precisely and allows for early - compile time - detection of incorrect handling of the dictionary values.
